akka-http-tests/src/test/scala/akka/http/scaladsl/server/SizeLimitSpec.scala

```diff
@@ -0,0 +1,248 @@
+        }
+      }
+    }
+
+    val binding = Http().bindAndHandle(route, "localhost", port = 0).futureValue
+
+    "reject a small request that decodes into a large chunked entity" in {
+      val request = Post(s"http:/${binding.localAddress}/noDirective", "x").withHeaders(`Content-Encoding`(HttpEncoding("custom")))
+      val response = Http().singleRequest(request).futureValue
+      response.status shouldEqual StatusCodes.BadRequest
+    }
+  }
+
+  "a route with decodeRequest that results in a large non-chunked streaming entity" should {
+    val decoder = decodeTo(nonChunkedEntityOfSize(decodeMaxSize + 1))
+
+    val route = path("noDirective") {
+      decodeRequestWith(decoder) {
+        post {
+          entity(as[String]) { e ⇒
+            complete(HttpEntity(ContentTypes.`text/html(UTF-8)`, s"Got request with entity of ${e.length} characters"))
+          }
+        }
+      }
+    }
+
+    val binding = Http().bindAndHandle(route, "localhost", port = 0).futureValue
+
+    "reject a small request that decodes into a large non-chunked streaming entity" in {
+      val request = Post(s"http:/${binding.localAddress}/noDirective", "x").withHeaders(`Content-Encoding`(HttpEncoding("custom")))
+      val response = Http().singleRequest(request).futureValue
+      response.status shouldEqual StatusCodes.BadRequest
+    }
+  }
+
+  "a route with decodeRequest followed by withoutSizeLimit" should {
+    val route = path("noDirective") {
+      decodeRequest {
+        withoutSizeLimit {
+          post {
+            entity(as[String]) { e ⇒
+              complete(HttpEntity(ContentTypes.`text/html(UTF-8)`, s"Got request with entity of ${e.length} characters"))
+            }
+          }
+        }
+      }
+    }
+
+    val binding = Http().bindAndHandle(route, "localhost", port = 0).futureValue
+
+    "accept a small request" in {
+      Http().singleRequest(Post(s"http:/${binding.localAddress}/noDirective", entityOfSize(maxContentLength)))
+        .futureValue.status shouldEqual StatusCodes.OK
+    }
+
+    "accept a small request that decodes into a large entity" in {
+      val data = ByteString.fromString("0" * (decodeMaxSize + 1))
+      val zippedData = Gzip.encode(data)
+      val request = HttpRequest(
+        HttpMethods.POST,
+        s"http:/${binding.localAddress}/noDirective",
+        immutable.Seq(`Content-Encoding`(HttpEncodings.gzip)),
+        HttpEntity(ContentTypes.`text/plain(UTF-8)`, zippedData))
+
+      zippedData.size should be <= maxContentLength
+      data.size should be > decodeMaxSize
+
+      val response = Http().singleRequest(request).futureValue
+      response.status shouldEqual StatusCodes.OK
+      response.entity.dataBytes.runReduce(_ ++ _).futureValue.utf8String shouldEqual (s"Got request with entity of ${decodeMaxSize + 1} characters")
+    }
+
+    // This is not entirely obvious: the 'withoutSizeLimit' inside the decodeRequest
+    // will also reset the size limit outside the decodeRequest.
```

Comment:
I find the comment is a bit misleading. If you use `withoutSizeLimit`, you basically waive your guarantees, and that works correctly and only inside the directive.

The non-obvious / counter-intuitive behavior might be that if decoding actually makes data smaller, then applying a tighter limit that would be passed after decoding could still catch on the original bigger data. This would probably be rare but when it happens it might be difficult to debug, because you may not find out where the final limit came from.